types/mongoose/index.d.ts

```diff
@@ -448,6 +448,8 @@ declare module "mongoose" {
 
     /** Expose the possible connection states. */
     static STATES: any;
+    then(res?: any, rej?: any): PromiseLike<any>;
+    catch(rej?: any): PromiseLike<any>;
```

Comment:
Thanks for making this update @isLishude! This PR is getting close now. 

The last thing to do, I think, is to use the built-in definitions for `Promise.then` and `Promise.catch` so that the arguments and return values to `then`/`catch` have more precise types. Specifically, I think they should be:

```
then: Promise<Connection>["then"];
catch: Promise<Connection>["catch"];
```

Then, in the tests, instead of doing:
```
mongoose.createConnection(connectUri).then(()=>{}, () => {});
```

you can do:

```
mongoose.createConnection(connectUri).then((conn)=> { conn.models; }, () => {});
```

Trying to access `conn.models` should implicitly verify that `conn` is correctly getting typed as a `Connection`.